In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from pandas_profiling import ProfileReport
import numpy as np

In [9]:
df = pd.read_csv('Advertising.csv')
df.tail()

Unnamed: 0     TV  radio  newspaper  sales
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

Problem statement 
1. build a modal that will predict the sales when the TV , radio and Newspaper advertisment is given in the dataset
2. if we are investing in TV ad or radio ad or newspaper then which mode of advertisment will be more effective
3. if there is relation between the TV ad, radio ad, and newspaper ad then show the relation between them(i.e linear relation or non relation)

Finally build the modal and do the deployment so that prediction can be made 

Also check the all the requirements for the Linear Regression

### At first we are looking for the statistic for the given data set

Here if we call the pandas profiling then it'll generate almost all the required statistics that we want to observe.

In [7]:
pf = ProfileReport(df)

in order to see the above result in more interactive way then we use .to_widgets() function

In [8]:
pf.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

Observation from the above report
1. check Alerts tab to know the feature is corelated to each other
2. in the variable tab, we can checck the details for each features mentioned. In a toggle detail option we can check the report of eahc feature in a more detailed way.
3. in the interaction tab, we can how one feature is corelated with the other feature. 
3.a here Tv with sales is linearly related i.e if we are going to increase the TV ad cost then sales will going to be increased.
3.b Incase of radio with sales there is linear relation
3.c incase of newspaper with sales there in no linear realtion been observed because the data points are scattered.
4. Now check the multicollinearity(i.e realtion among the dependent variable itself).
4.a There is no multicollinearity between tv and radio as the data points are shows random points
4.b There is no multicollinearity between tv and newspaper as the data points are shows random points

How to deal with multicollinear data?
We need drop such data because if we train our modal then in that case model will try to find out the relation among the feature itself. instead the modal needs to find the relation between the feature and the label(i.e x and y)


In [10]:
#save the report 
pf.to_file('Advertising_Report.html')

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

In [11]:
# here we are trying to build a modal between tv and sales
#i.e y= sales (because sales is the dependent variable and because of the changes in independent variable (tv ad) the sales is going to be affected or changed)
# and, x= tv advertisment cost
# we know the relation y= mx + c
x= df[["TV"]]
y= df[["sales"]]


In [14]:
# creating a  linear regression modal using sckit learn library
from sklearn.linear_model import LinearRegression
linear = LinearRegression() # creating object of the linear regression class
linear.fit(x,y) # inside this LinearRegression() , there are multiple function written . Here i am calling, fit() function so that I can do the training of the model(inside the fit() method complete training procedure is written down )
# whenever we write fit() method then it mean that our model is ready i.e final value of m and c is now calculated by the python


LinearRegression()

In [15]:
# showing the value of c
linear.intercept_

array([7.03259355])

In [16]:
#showing the value of m
linear.coef_

array([[0.04753664]])

### hence the equation can now be seen as ; 
### y=mx+c
### sales = 0.0475(TV) + 7.03259
Now if we put the TV ad cost then we can predict the sales generated due to the TV advertisment


In [19]:
#  now predict the sales based on the model that we have created by passing the Tv ad cost values(i.e in y=mx+c where value of m and c is computed by creating the linear model we are passing the value of x so that y can be calculated)
#here we are using the linear object to predict the sales
linear.predict([[45]])

E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[9.17174237]])

#### Therefore , if the Tv ad cost is 45 dollar then predicted sales will be 9.17174 dollar 

In [21]:
# passing multiple values of the Tv ad cost to find out the sales of the respective costs
l = [4,5,6,7,89,34,45,67,23]
for i in l :
    print(linear.predict([[i]]))

[[7.22274011]]
[[7.27027675]]
[[7.31781339]]
[[7.36535003]]
[[11.26335455]]
[[8.64883932]]
[[9.17174237]]
[[10.21754846]]
[[8.12593628]]


E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature nam

In [23]:
# here we are using model that we have saved using pickle to do prediction
# now save the above model in the .sav format(.sav is simple byte code format)
file_name = 'linear_reg.sav'
pickle.dump(linear, open(file_name, "wb"))
saved_model = pickle.load(open(file_name, 'rb'))
saved_model.predict([[45]])

E:\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[9.17174237]])

In [24]:
# finding out the accuracy of the model.
linear.score(x,y)

0.611875050850071

#### note : that in reality we never try to check the accuarcy with same dataset with which we have trained the model, instead we try to give the sepearate data set to find out the score 


## From Here Part 2 of the iNeuron Lecture is getting started

Here we are trying to create multi linear regression model (multi linear regression model says that when there are more than one feature is being involved (and one label data set involved) in the training of the model)


In the above scenario , we have taken only one single feature and one label and then created the model.
Now we ll take multiple feature and then create the model 



In [29]:
data_x = df[['TV', 'radio', 'newspaper']] # 3 Feature selected 
data_y = df[['sales']] # 1 label selected 

In [33]:
linear_1 = LinearRegression() # calling the linear regression class 
linear_1.fit(data_x, data_y) # training the model
print("value of c :", linear_1.intercept_) 
print("value of m :", linear_1.coef_)

value of c : [2.93888937]
value of m : [[ 0.04576465  0.18853002 -0.00103749]]


For the above different values of "m" we can interpret the equation as :

    y = m1(TV) + m2(radio) + m3(newspaper) + c
    
    y= 0.045(TV) + 0.188(radio) - 0.001(newspaper) + 2.938
    
Therefore,  when the advertisment expense of TV or radio or newspaper is given we can find its impact on the sales or how much each advertisment type is contributing to the sales.

In [34]:
#Now finding out the confindence score or the accuracy or the R square or the score 
linear_1.score(data_x, data_y)

0.8972106381789522

Observation: When we are considering all the 3 feature then the confidence score is approx 89.72 %

and when using only 1 column then the confidence score was approx 61.18 %

Hence we can say that we here we need to consider all the column in order to increase the accuracy of the model.(Lets see whether this statement is still correct untill the further examination)

In [36]:
# Lets try to find out the confidence score when we are using 2 feature (here , TV and radio)
data_x = df[['TV', 'radio']]
data_y = df[['sales']]
linear_2 = LinearRegression()
linear_2.fit(data_x, data_y)
print(linear_2.score(data_x, data_y))

0.8971942610828956


Observation : with respect to tv and radio the confidence score of the model is approx 89.71 %

Note: with respect to tv, radio, newspaper the score was 89.72%
 and with respect to tv, radio the score was 89.71%. We have not received much loss when we have excluded the newspaper feature. 
 We can also conclude that newspaper is not participating that much in the confidence score. Hence it can said that we are not suppose to invest much in the Newspaper advertisment.
 
Note: This is where Feature Selection comes into picture. Feature selection is selection of those feature with which we can build a model. Hence we do impact analysis of the features and then do Feature selection.

Note: one of the way for Feature selection is by comparing the R square.(R square is not the reliable technique of the Feature Selection)

Here in the above scenario we are comparing the score which is nothing but the R square only.

Note: the other way with which we can do Feature selection is by comparing the Adjusted R square of the individual features.

In [39]:
#Here we are using another approach of statistical modeling. We ll use Scikit learn library to build this model
import statsmodels.formula.api as smf
lm = smf.ols(formula = 'sales ~ TV ', data= df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Fri, 22 Jul 2022   Prob (F-statistic):           1.47e-42
Time:                        23:20:54   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Note : There are 2 ways of nuilding Linear Regression model
    
    Method1 : by using Scikit learn library
                
    Method2 : by using ols approach(ordinary least squared approach)

This is just the way to do impact analysis of the model that we trying to build or to decide which feature to choose to build a model

In [41]:
# using TV + radio for impact analysis
lm = smf.ols(formula = 'sales ~ TV + radio ', data= df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Fri, 22 Jul 2022   Prob (F-statistic):           4.83e-98
Time:                        23:26:29   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [43]:
# using TV + radio + newspaper for impact analysis
lm = smf.ols(formula = 'sales ~ TV + radio + newspaper ', data= df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Fri, 22 Jul 2022   Prob (F-statistic):           1.58e-96
Time:                        23:27:32   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
radio          0.1885      0.009     21.893      0.000       0.172       0.206
newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Note : the difference between Scikit Learn model and the statistical model(ols model) is that statistical model is not going to give the final version of the model which we can productionize. S
    statistical model(ols model) is just going to provide the summary and from that summary we just going to do our analysis and gain some insight for Feature selection

Observation with respect to the above 3 steps(i.e 1:- Tv and sales 
                                             2: Tv+radio and sales
                                             3: Tv+radio+newspaper and sales)
Now we are comparing the P values of all these to do impact analysis.

For Tv and Sales: p values is 0.0

For Tv + radio and sales: 0.0

For Tv + radio + newspaper: here for newspaper p value is 0.86 which means that 86 % chance is that null hypothesis is true and 14% chance is that alternate hypothesis is true.
                            i.e newspaper has only 14% impact on the sales which is negligible impact and hence it can be decided that newspaper Feature can dropped before trainig the model to decrease the performance and complexity of the model. 
                            
(p values means the probablity of null hypothesis being true.
Here, null hypothesis = newspaper ad has no impact on the sales which inturn says that they both are same and equal.
      Alternate hypothesis = newspaper ad has impact on the sales )

In [45]:
# The above is the Feature Selection procedure 